ADApp/ADSrc/NDAttribute.cpp

```diff
@@ -388,4 +388,25 @@ int NDAttribute::report(FILE *fp, int details)
   return ND_SUCCESS;
 }
 
+/** Get the max number of bits per pixel for the specified data type
+  * \param[in] NDDataType_t  Data type
+  */
+int GetNDDataTypeBits( NDDataType_t tyData )
+{
+    int nBits;
+    switch ( tyData )
+    {
+    // Default to large number so conversions to
+    // integer data types will be clipped to max
+    default:        nBits   = 99;   break;
+    case NDInt8:    nBits   = 8;    break;
+    case NDUInt8:   nBits   = 8;    break;
+    case NDInt16:   nBits   = 16;   break;
+    case NDUInt16:  nBits   = 16;   break;
+    case NDInt32:   nBits   = 32;   break;
+    case NDUInt32:  nBits   = 32;   break;
+    }
```

Comment:
GetNDDataTypeBits() is used to determine the default value for newly alloced NDArray's and for clipping in NDArrayPool::convert when converting to a smaller data type.   I agree it would be good to return -1 for invalid and specific values for NDFloat32 and NDFloat64.   I'm also open to returning a different value than 99 as long as it's large.   I don't think bitsPerPixel is particularly relevant for floating point NDArrays.

For color images, my searches seemed to indicate that discussions re the number of bits refer to the number of bits per pixel color, so 8 for typical RGB and 16 for RGB16.   I'm not aware of any color image standards that use anything but 8 or 16 except for digital video where you may see more bits for luminance, for example, 6 bits G, 5 bits R and 5 bits B.   For color, the
bitsPerPixel parameter is of less interest as the color mode tells you all you need to know.

Where bitsPerPixel is needed is for mono images, where there is no standard.   We have mono cameras with 8, 10, 12, and 16 bits and sometimes use the colorConvert plugin to reduce waveform size by converting to Uint8 images.

To be useful, BitsPerPixel must be set by the driver.   I've done this in one of my ADProsilica pull requests and in the ADEdtPdv driver I'll soon be posting for EDT framegrabber based cameras.